### PR TITLE
Type acquisition support for bower_components directory

### LINKED ISSUE #13917: [Type Acquisition] auto d.ts acquisition no longer works for bower_components
**TypeScript Version:** nightly 

Noticed the following issue in VS.  I'm filing this issue on the script side as the logic to determine the d.ts acquisition from bower_component occurs here.  (I don't believe there is anything on the managed side that comes into play)

1.	File > New Project > ASP.NET Web Application > Empty
2.	Add a JS file
3.	Right-click on project in solution explorer and  “Manage Bower Packages…”
4.	Install jquery
5.	Edit JS file

**Expected behavior:**

Expect "$. " to show completions 

**Actual behavior:**

jquery.d.ts is not included in the context

jquery

### PR DESCRIPTION
**Pull request by jramsay**
Fixes #13917 [Type Acquisition] auto d.ts acquisition no longer works for bower_components 

Issue: retrieving typing names from the bower_components directory is not currently supported. We only retrieve typings for bower components when a bower.json is present. By default in VS, ASP.NET projects do not create a bower.json when using the _Manage Bower Packages_ option.  No typings are currently being acquired for these components. 
  
Fix: 
1. Apply getTypingNamesFromPackagesFolder that is already being used to retrieve typings for the  node_modules  folder to the bower_components folder. 
2. Include the packagesFolder(s) in the list of filesToWatch.

### CODE REVIEW

**Comment by jramsay** [comment]
//cc: @vladima , @mhegazy , @zhengbli , @billti

File: src/services/jsTyping.ts [hunk 1]
```diff
@@ -199,17 +202,20 @@ namespace ts.JsTyping {
...
         /**
-         * Infer typing names from node_module folder
-         * @param nodeModulesPath is the path to the "node_modules" folder
+         * Infer typing names from packages folder (ex: node_module, bower_components)
+         * @param packagesFolderPath is the path to the packages folder
+
          */
-        function getTypingNamesFromNodeModuleFolder(nodeModulesPath: string) {
+        function getTypingNamesFromPackagesFolder(packagesFolderPath: string) {
+            filesToWatch.push(packagesFolderPath);
```
Comment (zhengbli) [inline src/services/jsTyping.ts]: should we watch the path after confirming it exists?

File: src/harness/unittests/typingsInstaller.ts [hunk 2]
```diff
@@ -692,7 +692,53 @@ namespace ts.projectSystem {
...
+            projectService.openClientFile(app.path);
+
+            checkNumberOfProjects(projectService, { configuredProjects: 1 });
+            const p = projectService.configuredProjects[0];
+            checkProjectActualFiles(p, [app.path]);
+
+            installer.installAll(/*expectedCount*/ 1);
+
+            checkNumberOfProjects(projectService, { configuredProjects: 1 });
+            checkProjectActualFiles(p, [app.path, jqueryDTS.path]);
```
Comment (zhengbli) [inline src/harness/unittests/typingsInstaller.ts]: Can you also check what paths are being watched?

File: src/services/jsTyping.ts [hunk 3]
```diff
@@ -199,17 +202,20 @@ namespace ts.JsTyping {
...
+            if (!host.directoryExists(packagesFolderPath)) {
                 return;
             }
 
             const typingNames: string[] = [];
-            const fileNames = host.readDirectory(nodeModulesPath, [".json"], /*excludes*/ undefined, /*includes*/ undefined, /*depth*/ 2);
+            const fileNames = host.readDirectory(packagesFolderPath, [".json"], /*excludes*/ undefined, /*includes*/ undefined, /*depth*/ 2);
             for (const fileName of fileNames) {
                 const normalizedFileName = normalizePath(fileName);
                 if (getBaseFileName(normalizedFileName) !== "package.json") {
```
Comment (zhengbli) [inline src/services/jsTyping.ts]: For bower, should be checking `bower.json` instead

Comment (zhengbli) [inline src/services/jsTyping.ts]: Also, line 228 and below have some logics specifically targetting npm 3, now we should check if the json file name is `package.json` before applying these logics.

File: src/services/jsTyping.ts (hunk 1 above)
Comment (jramsay) [inline src/services/jsTyping.ts]: I kept this before the path existence check as we want to watch these paths even if they haven't been created yet. (ex: User creates an empty ASP.NET project, adds a js file then uses _Manage Bower Packages_ to install jQuery.)

File: src/harness/unittests/typingsInstaller.ts (hunk 2 above)
Comment (jramsay) [inline src/harness/unittests/typingsInstaller.ts]: Good idea - will add this to the test

File: src/services/jsTyping.ts (hunk 3 above)
Comment (jramsay) [inline src/services/jsTyping.ts]: We already retrieve the bower.json typings using getTypingsFromJson(bowerJsonPath, filesToWatch).  This addresses the case when there is no bower.json present and just a bower_components folder that contains the packages.  From my tests the bower_components folder is similar to the node_modules folder.  It contains a folder directory for each package with a package.json (bower_components/jQuery/package.json).  I've been testing using the ASP.NET UI.  Let me try bower cmd line and see if this is consistent.

**Comment by zhengbli** [comment]
👍

**Pull request merged**